### PR TITLE
pool: track pending cost per payer

### PR DESCRIPTION
**Pull request by libotony**
# Description

Addresses vechain/protocol-board-repo#311

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] New and existing E2E tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have not added any vulnerable dependencies to my code

### CODE REVIEW

File: txpool/tx_object_map.go [hunk 1]
```diff
@@ -45,26 +51,53 @@ func (m *txObjectMap) Add(txObj *txObject, limitPerAccount int) error {
...
-		m.quota[*d]++
 	}
 
-	m.quota[txObj.Origin()]++
+	var (
+		cost  *big.Int
+		payer thor.Address
+	)
+
+	if txObj.Cost() != nil {
```
Comment (otherview) [inline txpool/tx_object_map.go:71]: Is it possible to flood the mempool with no-cost transactions ?

File: txpool/tx_object_map.go [hunk 2]
```diff
@@ -73,17 +106,29 @@ func (m *txObjectMap) RemoveByHash(txHash thor.Bytes32) bool {
...
+			delegator := delegatorKey(*d)
+			if m.quota[delegator] > 1 {
+				m.quota[delegator]--
 			} else {
-				delete(m.quota, *d)
+				delete(m.quota, delegator)
+			}
+		}
+
+		if payer := txObj.Payer(); payer != nil {
```
Comment (otherview) [inline txpool/tx_object_map.go:125]:
```suggestion
		// update the pending list and current map of payers
		if payer := txObj.Payer(); payer != nil {
```

File: txpool/tx_object_map.go [hunk 3]
```diff
@@ -94,6 +139,17 @@ func (m *txObjectMap) RemoveByHash(txHash thor.Bytes32) bool {
 	return false
 }
 
+func (m *txObjectMap) UpdateCost(txObj *txObject) {
```
Comment (otherview) [inline txpool/tx_object_map.go:142]: Minor sugestion:
```suggestion
func (m *txObjectMap) UpdatePendingCost(txObj *txObject) {
```

File: txpool/tx_object_map.go [hunk 4]
```diff
@@ -45,26 +51,53 @@ func (m *txObjectMap) Add(txObj *txObject, limitPerAccount int) error {
...
+
+	m.quota[origin]++
 	m.mapByHash[hash] = txObj
 	m.mapByID[txObj.ID()] = txObj
+	if delegator != nil {
+		m.quota[*delegator]++
+	}
+	if cost != nil {
+		m.cost[payer] = cost
+	}
```
Comment (otherview) [inline txpool/tx_object_map.go:94]: Minor suggestion :

```suggestion
	m.quota[origin]++
	if delegator != nil {
		m.quota[*delegator]++
	}
	
	m.mapByHash[hash] = txObj
	m.mapByID[txObj.ID()] = txObj
	
	if cost != nil {
		m.cost[payer] = cost
	}
```

**Review by otherview — COMMENTED** [review]
Good stuff. It looks good to me, only that concern around possible no-cost txs ?

File: txpool/tx_object_map.go [hunk 5]
```diff
@@ -45,26 +51,53 @@ func (m *txObjectMap) Add(txObj *txObject, limitPerAccount int) error {
 		return nil
 	}
 
-	if m.quota[txObj.Origin()] >= limitPerAccount {
+	origin := originKey(txObj.Origin())
```
[the remark this answers is not in the document]
Comment (libotony) [reply txpool/tx_object_map.go:54]: I think the last three questions are related.
> Are these types used ?
> What if quota map[interface{}]int was quota map[thor.Address]int ?
> What if there was no originKey cast here ?

The two new types are designed to handle the situation for one account both being transaction origin and transaction delegator.

File: txpool/tx_object_map.go (hunk 1 above)
Comment (libotony) [reply txpool/tx_object_map.go:71]: Not possible at peer-to-peer level, possible at API level, won't propagate until executable.

File: runtime/resolved_tx.go [hunk 6]
```diff
@@ -113,19 +114,20 @@ func (r *ResolvedTransaction) BuyGas(state *state.State, blockTime uint64) (
...
 	}
 
-	prepaid := new(big.Int).Mul(new(big.Int).SetUint64(r.tx.Gas()), gasPrice)
+	// prepaid is the max total of gas cost available to spend on this transaction
+	prepaid = new(big.Int).Mul(new(big.Int).SetUint64(r.tx.Gas()), gasPrice)
 	if r.Delegator != nil {
 		var sufficient bool
 		if sufficient, err = energy.Sub(*r.Delegator, prepaid); err != nil {
 			return
 		}
```
Comment (otherview) [inline runtime/resolved_tx.go:123]: What do you think about this empty returns ? 👀

File: txpool/tx_object_map.go (hunk 1 above)
Comment (otherview) [reply txpool/tx_object_map.go:67]: It seems like they wouldn't be added until executable right ? So we should be good.

However theres seems to be a different behaviour [here](https://github.com/vechain/thor/blob/cc2b3764839ce910d979cc6d1c283d60acfc17e6/txpool/tx_pool.go#L279-L294). Let's pair up and review it separately to this PR tomorrow perhaps ?

**Review by otherview — APPROVED** [review]

**Pull request merged**